### DISCUSSION TITLE
Running in container

### DISCUSSION THREAD

**Discussion by Trophime** [Kitware/trame#43] [Show and tell]
Hi,
I would like to create a singularity container for running trame apps.
I've managed to get all examples from the tutorial running with this recipe:

```
Bootstrap: docker
From: ubuntu:focal
#From: DOCKERIMAGE:DOCKERTAG


%environment
export PATH=/opt/ParaView-5.10.0-MPI-Linux-Python3.9-x86_64/bin:$PATH
export LANG=en_US.UTF-8
export LANGUAGE=en_US.UTF-8
export LC_ALL=C.UTF-8
export OMPI_MCA_btl_vader_single_copy_mechanism=none
export PYTHONPATH=/usr/local/lib/python3.9/dist-packages:$PYTHONPATH

PORT=1234
ALLOW_HTTP=true
URL=localhost
export PORT ALLOW_HTTP URL
    
%labels
AUTHOR christophe.trophime@lncmi.cnrs.fr
Maintainer: Christophe Trophime
Maintainer_email: christophe.trophime@lncmi.cnrs.fr
Version 1.0

%post

# install 
apt-get -qq update
apt-get -y install lsb-release sudo apt-file locate tree
apt-get -y install curl wget
DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends tzdata git
DEBIAN_FRONTEND=noninteractive apt-get -y install python-is-python3 python3-pip
DEBIAN_FRONTEND=noninteractive apt-get -y install python3.9-minimal python3.9-distutils python3.9-venv libpython3.9-dev
apt-get -y install openmpi-bin openmpi-common

# Required to run ParaView
apt-get install -y --no-install-recommends libxrender1 libxcursor1 libsm6 libxkbcommon-x11-0 fontconfig

apt-file update
updatedb

# Install Paraview
wget "https://www.paraview.org/paraview-downloads/download.php?submit=Download&version=v5.10&type=binary&os=Linux&downloadFile=ParaView-5.10.0-MPI-Linux-Python3.9-x86_64
.tar.gz" -O /tmp/ParaView-5.10.0-MPI-Linux-Python3.9-x86_64.tar.gz
tar zxvf /tmp/ParaView-5.10.0-MPI-Linux-Python3.9-x86_64.tar.gz
mkdir -p /opt
mv ParaView-5.10.0-MPI-Linux-Python3.9-x86_64 /opt
rm -rf /tmp/ParaView-5.10.0-MPI-Linux-Python3.9-x86_64.tar.gz

# Install Trame
update-alternatives --install /usr/bin/python python /usr/bin/python3.9 1
python3.9 -m pip install "trame"
python3.9 -m pip install "vtk>=9.1.0"
```   
The demos are accessible launching them from the command line and running a web browser using http://localhost:1234/ or http://localhost:8080/ 

The only trouble is that when running paraview examples, it launches beside the web page another window with the model.
If you find how to improve this recipe, I would be glad to hear from you.

I wish to create a service to provide visualisation for FEM calculation.
I would need to call the trame app from a frontend ad pass for instance the name of the file vtu or pvsm to be displayed.
How can I pass the filename from the frontend to the trame app?

I wouldalso  like to know what I need to do to share this service on a multi-user system.
In the output from trame app:
"Note that for multi-users you need to use and configure a launcher."

Could you please provide an example of running on such a system? or points me to some docs/ex?

Best
C

PS: with docker I cannot run examples requiring opengl 4.xx even when using as a base image nvidia/opengl:1.2-glvnd-runtime.

**Comment by jourdain** [comment]
Thanks @Trophime for sharing your singularity receipt. 

For supporting remote rendering, we should be using an EGL bundle. VTK is still missing its bundle with EGL. But ParaView can be downloaded for EGL which also means it can be use as a VTK runtime. 

If you want to look at a multi-user example with docker and remote rendering, you can look at what I've done for the next visualizer [here](https://github.com/Kitware/paraview-visualizer/tree/master/docker).

The only trick for configuring your app inside docker, is that you [need to be explicit for running your app](https://github.com/Kitware/paraview-visualizer/blob/master/docker/setup/apps.yml#L4-L19) with paraview/python.

Then if you want to provide additional arguments that are user specific (from url), you can create other application targets like I've done [here](https://github.com/Kitware/paraview-visualizer/blob/master/docker/setup/apps.yml#L20-L36) where we expect a `&data=/path/to/file/browsing` arg in the URL. Also the `name=xxx` is mandatory otherwise none of the url param will be provided at startup.

**Reply by moon-chyldren** [reply]
@jourdain It looks like the 3 links in this reply are now dead and give me a 404. Do you know where I could locate the information you provided in these links?

**Reply by jourdain** [reply]
Yes we made visualizer private for now as we couldn't get funding to maintain it. 

Feel free to post a new discussion on what ever topic you are looking to solve. A lot has changed on the rendering backend topic. Now both VTK (9.4+) and ParaView (v6) support all the backend from a single binary.

**Comment by jourdain** [comment]
A wheel of `vtk-egl` is now available via our Kitware registry.

`pip install --extra-index-url https://wheels.vtk.org vtk-egl`
